saleor/graphql/discount/mutations/sale/sale_channel_listing_update.py

```diff
@@ -82,61 +75,76 @@ class Meta:
         error_type_field = "discount_errors"
 
     @classmethod
-    def add_channels(cls, promotion: Promotion, add_channels: List[Dict]):
-        rules_data_to_add: List[RuleInfo] = []
-        rules_to_update: List[PromotionRule] = []
-
-        rules = promotion.rules.all()
-        PromotionRuleChannel = PromotionRule.channels.through
-        rule_channel = PromotionRuleChannel.objects.filter(
-            Exists(rules.filter(id=OuterRef("promotionrule_id")))
-        )
-        channelid_ruleid_map = {
-            channel_id: rule_id
-            for channel_id, rule_id in rule_channel.values_list(
-                "channel_id", "promotionrule_id"
-            )
-        }
-        examplary_rule = rules[0]
+    def add_channels(
+        cls,
+        promotion: Promotion,
+        exemplary_rule: PromotionRule,
+        add_channels: List[Dict],
+    ):
+        channel_id_rule_map = cls.get_channe_id_to_rule_map(promotion)
+        rules_to_create: List[Tuple[Channel, PromotionRule]] = []
+        rules_to_update: List[Tuple[Channel, PromotionRule]] = []
         for add_channel in add_channels:
             channel = add_channel["channel"]
             discount_value = add_channel["discount_value"]
 
-            if channel.id not in channelid_ruleid_map.keys():
-                rules_data_to_add.append(
-                    RuleInfo(
-                        rule=PromotionRule(
+            if channel.id not in channel_id_rule_map and exemplary_rule:
```

Comment:
What if `exemplary_rule` does not exist? It will crash in `else` part

Reply:
We ensure, that promotion has at least on rule. In fact, this check is redundant.